src/thor/timedep_forward.cc

```diff
@@ -0,0 +1,285 @@
+
+constexpr uint64_t kInitialEdgeLabelCount = 500000;
+
+// Default constructor
+TimeDepForward::TimeDepForward()
+    : AStarPathAlgorithm() {
+  mode_ = TravelMode::kDrive;
+  travel_type_ = 0;
+  adjacencylist_ = nullptr;
+  max_label_count_ = std::numeric_limits<uint32_t>::max();
+}
+
+// Destructor
+TimeDepForward::~TimeDepForward() {
+  Clear();
+}
+
+// Expand from the node along the forward search path. Immediately expands
+// from the end node of any transition edge (so no transition edges are added
+// to the adjacency list or EdgeLabel list). Does not expand transition
+// edges if from_transition is false.
+void TimeDepForward::ExpandForward(GraphReader& graphreader,
+                   const GraphId& node, const EdgeLabel& pred,
+                   const uint32_t pred_idx, const bool from_transition,
+                   const uint32_t localtime,
+                   const odin::Location& destination,
+                   std::pair<int32_t, float>& best_path) {
+  // Get the tile and the node info. Skip if tile is null (can happen
+  // with regional data sets) or if no access at the node.
+  const GraphTile* tile = graphreader.GetGraphTile(node);
+  if (tile == nullptr) {
+    return;
+  }
+  const NodeInfo* nodeinfo = tile->node(node);
+  if (!costing_->Allowed(nodeinfo)) {
+    return;
+  }
+
+  // Expand from end node.
+  uint32_t shortcuts = 0;
+  uint32_t max_shortcut_length = static_cast<uint32_t>(pred.distance() * 0.5f);
+  GraphId edgeid(node.tileid(), node.level(), nodeinfo->edge_index());
+  EdgeStatusInfo* es = edgestatus_.GetPtr(edgeid, tile);
+  const DirectedEdge* directededge = tile->directededge(nodeinfo->edge_index());
+  for (uint32_t i = 0; i < nodeinfo->edge_count(); ++i, ++directededge, ++edgeid, ++es) {
+    // Handle transition edges - expand from the end node of the transition
+    // (unless this is called from a transition).
+    if (directededge->trans_up()) {
+      if (!from_transition) {
+        hierarchy_limits_[node.level()].up_transition_count++;
+        ExpandForward(graphreader, directededge->endnode(), pred, pred_idx,
+                      true, localtime, destination, best_path);
+      }
+      continue;
+    } else if (directededge->trans_down()) {
+      if (!from_transition &&
+          !hierarchy_limits_[directededge->endnode().level()].StopExpanding(pred.distance())) {
+        ExpandForward(graphreader, directededge->endnode(), pred, pred_idx,
+                      true, localtime, destination, best_path);
+      }
+      continue;
+    }
+
+    // Skip this edge if permanently labeled (best path already found to this
+    // directed edge), if edge is superseded by a shortcut edge that was taken,
+    // or if no access is allowed to this edge (based on costing method), or if
+    // a complex restriction exists.
+    if (es->set() == EdgeSet::kPermanent ||
+        (shortcuts & directededge->superseded()) ||
+       !costing_->Allowed(directededge, pred, tile, edgeid, localtime) ||
```

Comment:
Should this be something like `localtime + std::accumulate(best_path.begin(), best_path.end(), [](auto &a, auto &b) { return a+b; });` so that restrictions in the future are considered further down the route?

Reply:
not sure I understand. localtime accounts for the elapsed time along the best path to this edge so it considers whether the edge is allowed (access restrictions in this case) based on any time dependent access restrictions.

Reply:
Ah, nevermind - I misunderstood that EdgeForward was a single edge relaxation step.  I can see the progression of `localtime` now in `GetBestPath`.